src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -47,145 +53,224 @@ class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[
-  def matches: Boolean = portRanges.isDefined
-
-  /**
-    * @return the resulting assigned ports.
-    */
-  def ports: Seq[Long] = portRanges.map(_.flatMap(_.ranges.flatMap(_.asScala()))).getOrElse(Nil)
-
-  /**
-    * Try to satisfy all app.ports exactly (not dynamically) if that's possible to do with one of the offered
-    * ranges.
+    * Try to find supplied ports in offer. Returns `None` if not all ports were found.
     */
-  // TODO use multiple ranges if one is not enough
-  private def appPortRanges: Option[RangesResource] = {
-    val sortedPorts = app.ports.sorted
-    val firstPort = sortedPorts.head
-    val lastPort = sortedPorts.last
-
-    lazy val matchingRange: Option[PortRange] = offeredPortRanges.find {
-      case PortRange(_, range) => range.getBegin <= firstPort && range.getEnd >= lastPort
-      case _                   => false
-    }
-
-    // Monotonically increasing ports
-    if (firstPort + sortedPorts.size - 1 == lastPort && matchingRange.isDefined) {
-      Some(RangesResource(
-        Resource.PORTS, Seq(protos.Range(firstPort.longValue, lastPort.longValue)),
-        matchingRange.get.role))
-    }
-    else {
-      val suitablePortRangeOpt = offeredPortRanges
-        .iterator
-        .find {
-          case PortRange(_, range) => app.ports.forall(p => range.getBegin <= p && range.getEnd >= p)
+  private[this] def findPortsInOffer(requiredPorts: Seq[Integer], failLog: Boolean): Option[Seq[PortWithRole]] = {
+    takeEnoughPortsOrNone(expectedSize = requiredPorts.size) {
+      requiredPorts.sorted.iterator.map { port =>
+        offeredPortRanges.find(_.range.contains(port)).map { offeredRange =>
+          PortWithRole(offeredRange.role, port)
+        } orElse {
+          if (failLog) log.info(s"Couldn't find host port $port in any offered range for app [${app.id}]")
+          None
         }
-      suitablePortRangeOpt.map {
-        case PortRange(role, range) =>
-          val portRanges = app.ports.map(p => protos.Range(p.longValue, p.longValue))
-          RangesResource(Resource.PORTS, portRanges, role)
       }
     }
   }
 
   /**
-    * @return dynamically assign ports if that's possible to do with one of the offered
-    * ranges.
+    * Choose random ports from offer.
     */
-  // TODO use multiple ranges if one is not enough
-  private def randomPortRanges: Option[RangesResource] = {
-    for (PortRange(role, range) <- offeredPortRanges) {
-      if (range.getEnd - range.getBegin + 1 >= app.ports.length) {
-        val maxOffset = (range.getEnd - range.getBegin - app.ports.length + 2).toInt
-        val firstPort = range.getBegin + Random.nextInt(maxOffset)
-        val rangeProto = protos.Range(firstPort, firstPort + app.ports.length - 1)
-        return Some(
-          RangesResource(Resource.PORTS, Seq(rangeProto), role)
-        )
-      }
+  private[this] def randomPorts(numberOfPorts: Int): Option[Seq[PortWithRole]] = {
+    takeEnoughPortsOrNone(expectedSize = numberOfPorts) {
```

Comment:
Summarizing the discussion. I see the following alternatives:

* `({...})` instead of `{...}`
* Making the parameter unnecessarily a function
* calculating the parameter before the method invocation and pass per variable name

I don't like anything better than the current choice.